File: TwoPointer/713. Subarray Product Less Than K/713. Subarray Product Less Than K.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        // fix the right boundary
        int n = nums.size();
        int prod = 1;
        int j = 0;
        int count = 0;
        for (int i = 0;  i < n; i++) {
            if (j < i) {
                j = i;
                prod = 1;
            }
            while (j < n && prod*nums[j] < k) {
                prod *= nums[j];
                j++;
            }
            count += j-i;
            prod /= nums[i];
        }
        return count;
    }
};

// X [X X X X X] X) X
//    i          j
   
// 10, *, *, * k = 5

File: TwoPointer/713. Subarray Product Less Than K/713. Subarray Product Less Than K.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int ret = 0;
        int i = 0;
        long sum = 1;
        for (int j = 0; j < nums.length; j++) {
            sum *= nums[j];
            int curleft = i;
            if (sum < k) ret += j-i+1;
            while (i < j && sum >= k) {
                sum /= nums[i++];
            }
            if (i > curleft && sum < k) ret += j-i+1;
        }
        return ret;
    }
}

/*class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1) return 0;
        // contiguous
        // two pointer
        // positive integer no zero
        int len = nums.length, left = 0, right = 0;
        int res = 0;
        long product = 1;
        //boolean first = true;
        while (right < len) {
            product *= nums[right];
            
            
            while (product >= k) {
                //if (first) res += (right-left+1)*(right-left)/2;
                //System.out.println("left idex -> " + left + " : right idex -> " + right);
                
                product /= nums[left];
                left++;
            }
            //if (product < k) res++;
            res += right-left+1;
            right++;
        }
        //System.out.println("last left idex -> " + left + " : last right idex -> " + right);
        //if (product < k) res += (right-left+1)*(right-left)/2;
        return res;
    }
}*/
